lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h

```diff
@@ -0,0 +1,83 @@
+
+#include "llvm/ADT/STLExtras.h"
+
+#include "swift/Reflection/ReflectionContext.h"
+
+
+namespace lldb_private {
+struct SwiftMetadataCache {
+public:
+  static SwiftMetadataCache &get(); 
+  SwiftMetadataCache(SwiftMetadataCache &) = delete;
+  void operator=(SwiftMetadataCache const&) = delete;
+
+  /// Relate the lldb module with the reflection info id.
+  void registerModuleWithReflectionInfoID(lldb::ModuleSP module,
+                                          uint64_t info_id);
+
+  /// Cache the field descriptors in the info with the mangled names
+  /// passed in. The number of field descriptors and mangled names passed should
+  /// be the same, otherwise caching is aborted.
+  void cacheFieldDescriptors(
+      uint64_t info_id,
+      const swift::reflection::FieldSection &field_descriptors,
+      const std::vector<std::string> &mangled_names);
+
+  llvm::Optional<std::pair<uint64_t, uint64_t>> 
+    getFieldDescriptor(const std::string &mangled_name) const;
+
+  /// Returns whether looking for a mangled name in reflection info with a given
+  /// id could possibly return information that wasn't already in the cache.
+  bool shouldParseReflectionInfoWithId(uint64_t info_id) const;
+
+private:
+  SwiftMetadataCache() = default;
+
+  void loadCacheForModule(lldb::ModuleSP module, uint64_t info_id);
+
+  bool writeMangledNamesAndOffsetsToEncoder(
+      uint64_t info_id,
+      const swift::reflection::FieldSection &field_descriptors,
+      const std::vector<std::string> &mangled_names, DataEncoder &encoder);
+
+  void fillMangledNameToOffsetMap(
+    uint64_t info_id, const swift::reflection::ReflectionInfo &info,
+    const std::vector<std::string> &mangled_names);
+
+  /// Gets the file name that the cache file should use for a given module.
+  std::string
+  getTyperefCacheFileNameForModule(const lldb::ModuleSP &module);
+
+  /// A map from mangled names to a pair constituting of the reflection info's
+  /// id and the offset of the field descriptor in the field section buffer it's
+  /// in.
+  llvm::StringMap<std::pair<uint64_t, uint64_t>> m_mangled_name_to_offset;
+
+  /// A map from reflection infos ids to a pair constituting of its
+  /// corresponding module and whether or not we've inserted the cached metadata
+  /// for that reflection info already.
+  std::unordered_map<uint64_t, std::pair<lldb::ModuleSP, bool>>
+      m_info_to_module;
+
+  /// Version number that indicates how the data is laid out.
+  /// Version 1 is encoded as follows:
+  /// A header constituting of:
+  /// - Version number (uint16_t).
+  /// - Module signature.
+  /// - Size of the remainder of the contents when decompressed (uint64_t).
+  /// A body that is zlib compressed containing:
+  /// - The number (N) of name/offset pairs (uint64_t).
+  /// - N pairs composed of a c-string mangled name followed by the offset
+  ///   (uint32_t) where the corresponding field descriptor can be found in
+  //    the fieldmd section.
```

Comment:
Are the pairs sorted for faster reading?

Reply:
With the current implementation I don't think that'd matter, as we parse all strings to build the map object anyway. I think we can address this in a second patch, maybe by changing the storage to use `OnDiskHashTable` instead. I didn't want to add more to this patch as it already seemed quite big.